src/vos/vos_obj.c

```diff
@@ -385,9 +385,13 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 		D_DEBUG(DB_IO, "Failed to punch object "DF_UOID": rc = %d\n",
 			DP_UOID(oid), rc);
 
-	if ((rc == -DER_NONEXIST || rc == 0) &&
-	    vos_ts_wcheck(ts_set, epr.epr_hi, bound))
-		rc = -DER_TX_RESTART;
+	if (rc == 0 || rc == -DER_NONEXIST) {
+		/** We must prevent underpunch with regular I/O */
+		if (rc == 0 && (flags & VOS_OF_REPLAY_PC) == 0)
```

Comment:
A bit worried... If VOS_OF_REPLAY_PC is set, we should skip all MVCC checks. Does this here implies that we are performing wcheck before for rebuild I/Os? Does it work because for rebuild I/Os, we must have bound == epoch?

Reply:
> A bit worried... If VOS_OF_REPLAY_PC is set, we should skip all MVCC checks. Does this here implies that we are performing wcheck before for rebuild I/Os? Does it work because for rebuild I/Os, we must have bound == epoch?

Yes, we always have bound == epoch for rebuild.   Note, I need to clean up this code in general but I'd like to leave that to future work at this point.

Reply:
Sure, just wanted to make sure I was following.